src/File.hack

```diff
@@ -22,6 +22,45 @@ final class File {
     return $this->isDirty;
   }
 
+  private function getLines(): vec<string> {
+    return Str\split($this->contents, "\n");
+  }
+
+  const type TLineNumber1Base = int;
+
+  /**
+   * Returns the 1-base line number of lint markers in the file.
+   */
+  <<__Memoize>>
+  public function lintMarkers(): dict<string, keyset<self::TLineNumber1Base>> {
+    return $this->getLines()
+      |> Vec\map_with_key(
+        $$,
+        ($line_index, $line_content) ==>
+          Regex\every_match($line_content, re"/(?<marker_name>\w+)\[\w+\]/")
+          |> Vec\filter(
+            $$,
+            $match ==> LintMarkerName::isValid($match['marker_name']),
+          )
+          |> Vec\map(
+            $$,
+            $match ==> shape(
+              'line_number' => $line_index + 1,
+              'marker' => $match[0],
```

Comment:
Should this really be returning the marker rather than just the code?

Reply:
We actually need two indices: the marker name and the code. I previously let it return a `dict<LintMarkerName, dict<TErrorCode, keyset<TLineNumber>>>`, which is roughly equivalent to the current return type `dict<string, keyset<TLineNumber1>>`. I changed to the current return type because it is closer to the existing way to look up a marker.

Reply:
Do we actually need the marker name, especially if we bring in #351? I'd expect us to just allow any form of suppresion for any code - e.g. both `HHAST_IGNORE_ERROR[1234]` and `@lint-ignore 1234`

Reply:
I expect the marker name would be configurable in some extent, e.g. whether support `@lint-ignore`, `HH_FIXME`, `HHAST_FIXME`, etc. 

Including the the marker name would allow the ability of configuration.

Reply:
- HH_* will definitely not be used
- I don't see a need for this to vary by linter; I can imagine some particular projects wanting to:
  - insist that all suppressions use the same format (in which case, that could be a linter, or hhast-wide configuration)
  - require 'these linters should use @lint-ignore, these should use <rule>' - but this seems unlikely
  
For both of the latter cases, the logic should likely be centralized rather than reproduced in each linter (or the base class)

Reply:
Another reason why the marker name information is needed is that I use this function to cache `HHAST_IGNORE_ALL` as well, which is treated differently from `HHAST_FIXME` and `HHAST_IGNORE_ERROR `.

Reply:
we should also support `@lint-ignore-all` in the future though; ultimately every linter should be:
- splitting the marker into marker name and error code
- figuring out if it's a -all marker
- acting appropriately

How about rename it to `getLintSuppressions`, and the return type changed to:

```hack
shape(
  'all' => keyset<CODE_AS_STRING>,
  'lines' => dict<int, keyset<CODE_AS_STRING>>
)
```

... though then the doc comment should emphasize that `lines` should only be used by line-based linters.

This (and ASTLinter) should share a common source of truth about which names are 'all' vs just-one-occurence

Reply:
I just updated this PR to separate file-level and line-level markers